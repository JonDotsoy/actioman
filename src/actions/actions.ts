import type { z } from "zod";
import type { Configs, ConfigsModule } from "../configs/configs.js";
import { get } from "@jondotsoy/utils-js/get";
import actionsList from "../share-actions/share-actions.js";

export const actions = actionsList;
const GeneratorFunction = function* () {}.constructor;
const AsyncGeneratorFunction = async function* () {}.constructor;

type InferZodType<T> = T extends z.ZodType ? z.infer<T> : any;
type ParameterTypeInferer<T> = T extends (i: infer R) => any ? R : any;
type Handler<
  InputType = any,
  ResultType = any,
  SSE extends boolean = any,
> = SSE extends true
  ? (
      input: InferZodType<InputType>,
    ) =>
      | Generator<InferZodType<ResultType>>
      | AsyncGenerator<InferZodType<ResultType>>
  : (
      input: InferZodType<InputType>,
    ) => InferZodType<ResultType> | Promise<InferZodType<ResultType>>;

export type Action<
  InputType = any,
  ResultType = any,
  SSE extends boolean = false,
  T = any,
> = {
  description?: string;
  input?: InputType;
  output?: ResultType;
  sse?: SSE;
  handler: Handler<InputType, ResultType, SSE>;
};

export const defineAction = <I = any, R = any, SSE extends boolean = false>(
  action: Action<I, R, SSE>,
): Action<I, R, SSE> => action;

export const actionFromModule = (module: any) => {
  const handlerFn = get.function(module) as
    | undefined
    | ((...args: any[]) => any);
  if (handlerFn) {
    const isGenerator =
      handlerFn instanceof GeneratorFunction ||
      handlerFn instanceof AsyncGeneratorFunction;
    return defineAction({ sse: isGenerator, handler: handlerFn });
  }
  const handlerObj = get.record(module);
  if (!handlerObj) return null;
  const handler = get.function(handlerObj, "handler") as
    | undefined
    | ((...args: any[]) => any);
  if (!handler) return null;
  return defineAction({
    description: get.string(handlerObj, "description"),
    input: get.record(handlerObj, "input"),
    output: get.record(handlerObj, "output"),
    handler,
  });
};

export class Actions<ActionsDefinitions extends Record<string, Action> = any> {
  constructor(private definition: ActionsDefinitions) {}

  /** @deprecated It has never been used */
  call = async <NameActionDefinition extends keyof ActionsDefinitions>(
    name: NameActionDefinition,
    input: ParameterTypeInferer<
      ActionsDefinitions[NameActionDefinition]["handler"]
    >,
  ): Promise<
    ReturnType<ActionsDefinitions[NameActionDefinition]["handler"]>
  > => {
    const definition = this.definition[name];
    if (!definition) throw new Error(`Action ${name.toString()} not found`);
    return await definition.handler(input);
  };

  static describe(actions: Actions): Record<string, Action> {
    return actions.definition;
  }

  static fromModule(module: unknown, configs?: ConfigsModule) {
    const isRecord =
      typeof module === "object" && module !== null && !Array.isArray(module);

    if (!isRecord) return null;

    const actionDefinitions: Record<string, Action<any, any, any>> = {};

    for (const [moduleName, handler] of Object.entries(module)) {
      const definition = actionFromModule(handler);
      if (!definition) break;
      actionDefinitions[moduleName] = definition;
    }

    return new Actions(actionDefinitions);
  }
}
