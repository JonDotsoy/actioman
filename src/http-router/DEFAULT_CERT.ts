export const DEFAULT_CERT =
  "-----BEGIN CERTIFICATE-----\n" +
  "MIIDCTCCAfGgAwIBAgIUcx4/uMJz+em1t9pyg8YBn8mOtz8wDQYJKoZIhvcNAQEL\n" +
  "BQAwFDESMBAGA1UEAwwJbG9jYWxob3N0MB4XDTI1MDIwNzE1MDcyMloXDTI1MDMw\n" +
  "OTE1MDcyMlowFDESMBAGA1UEAwwJbG9jYWxob3N0MIIBIjANBgkqhkiG9w0BAQEF\n" +
  "AAOCAQ8AMIIBCgKCAQEApcFu5mawUooMhsH/1bYdRMn1PTlmRiPRJkoookkWOhVH\n" +
  "703iYVCkZL39DF9tXG4iANNdjFepVIcSZduyZBXslAvuvFWaNu9I52IYyDYDadDT\n" +
  "oLvXFi3dsfnIrGf68JLBqZnOSxVLg80Ht3maOSbc36DbUO5YE9nU0nhA/wQ983j0\n" +
  "FDzBKzrPYPFNIbmI3SEVPjhrFrxci7+Tbdky4SpIwLRN7SWpbu/GPmLzlvD8XMuW\n" +
  "cw1dsBoqV9w+pMwy2DTctQ3SU3nbxuvMgMKu4wuX9mWLiCmykGcLvYUrI8CKFzzP\n" +
  "f48Qd2/HC1lzwCKiwn57MhokbzcKWhB8E96X3HGtVwIDAQABo1MwUTAdBgNVHQ4E\n" +
  "FgQUPsTW/85jbxCCvyiMwx33xM6nh/IwHwYDVR0jBBgwFoAUPsTW/85jbxCCvyiM\n" +
  "wx33xM6nh/IwDwYDVR0TAQH/BAUwAwEB/zANBgkqhkiG9w0BAQsFAAOCAQEANfYP\n" +
  "er5WrTFuWtY2ApxMjvDWo0xuL6+kCv4UkaLtfsR4TkcF9svGtb/PvrlOmffP+By1\n" +
  "TkBoMju3uHU7/hPo5UMRUI445QnIVSSpLaHnLHBi85KGMl55ErRbOGVgYR1n+Qxo\n" +
  "nFwucn2hjHBmswXNY9DJx8vmewKqa9Kc694YxNx44tmMNbokUYKliofnisxxhM3Y\n" +
  "4REA+w3bme7z7JDLt0J1EGCAEeT2NcLbnXw8gEKJ0D39Zt5b5qR1Re4wWtknXlkc\n" +
  "5VGw96c5jbejD49BJsDWqVsJG1D4dhJ61ybBT6j2aRnSH+MxPHpH/Rm9wXIGha4G\n" +
  "mZLuodbHDBrL/FT4cQ==\n" +
  "-----END CERTIFICATE-----\n";
