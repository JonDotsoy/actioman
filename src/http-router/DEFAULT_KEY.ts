if (process.env.NODE_ENV !== "test") {
  console.warn("The variable DEFAULT_KEY is only to test environment");
}

export const DEFAULT_KEY =
  "-----BEGIN PRIVATE KEY-----\n" +
  "MIIEvAIBADANBgkqhkiG9w0BAQEFAASCBKYwggSiAgEAAoIBAQClwW7mZrBSigyG\n" +
  "wf/Vth1EyfU9OWZGI9EmSiiiSRY6FUfvTeJhUKRkvf0MX21cbiIA012MV6lUhxJl\n" +
  "27JkFeyUC+68VZo270jnYhjINgNp0NOgu9cWLd2x+cisZ/rwksGpmc5LFUuDzQe3\n" +
  "eZo5JtzfoNtQ7lgT2dTSeED/BD3zePQUPMErOs9g8U0huYjdIRU+OGsWvFyLv5Nt\n" +
  "2TLhKkjAtE3tJalu78Y+YvOW8Pxcy5ZzDV2wGipX3D6kzDLYNNy1DdJTedvG68yA\n" +
  "wq7jC5f2ZYuIKbKQZwu9hSsjwIoXPM9/jxB3b8cLWXPAIqLCfnsyGiRvNwpaEHwT\n" +
  "3pfcca1XAgMBAAECggEAGE8voAvEZT6IysSt7B5a2YlS9aky5it/k4GBxxc38snH\n" +
  "NIH0JviIO5wffg6Kv+kJBJz7FcePdm2e1aBsX8BVY3sChs15IgBi7UhWQRqm0oCF\n" +
  "YQfbag3M+zqaYyqWw2yKiYCcoCdd3VUXAiIL68DW2NdDAvIJqdOjIkDvq1bobQn+\n" +
  "MNdNBnu58t95JRfEbQHiDUUVTGv1b4E8gAUtCmsDPtsSQAox8rTD4UTA0+9oxE5e\n" +
  "SoV69YWS+BqFaDZgxxTKxyoL6g4K7WrF58oOSIlayUBGOO0VpGBf3sFB9YvYw76P\n" +
  "7m8S1EY71G8/Bp4JoVrNR7TemG0Qotc/3AapEjyv7QKBgQDVKfOWe81oKuFJlpKP\n" +
  "1W0v7xoo0fV9GXJlOrpDTvcuoY+XiMuQtTWDl7de+4y+mzQ+YUWKCrEn6yj0eUgy\n" +
  "owEpk8p6E+nvU0w9Yhnlk0wmfHkQw/P4kwCuG7sFvcpJ8471Ad94NWLddeHdVQZb\n" +
  "3A8PZgXouoBstTlKJ9WkZHxmCwKBgQDHEJmiR/U6m882A8G7yhoDB4U2tYPS2WZq\n" +
  "sZdgaCDrYZNME46N11b6HEJYD6jHyP33FPgyLD9uz157/1WTod6y2fdrlRnTWXZt\n" +
  "KKX837tOaKZFYpJtTzVFdR6nCO6O4fL6tK/CyNIWOB+kAXmtr8GBfbBfGHuePuYr\n" +
  "9jjkGmIhZQKBgBes7Cgnof3S3Wjd93NQQitTxDUKdNirLeB+cFevepbMTPWItQti\n" +
  "hQoTnMNI7y3wcRIInb+iRmDnponYuqMdJbDngFKkyZPSTaJ3teagwJF5u7yX9w4B\n" +
  "zc6It/IATWp/Sb5+scJZM5zjTe3WVrl3LJkKPNfHdyIb0rwgxoLWwlZXAoGALI8s\n" +
  "tMTCwgHKml6w6SZOOEtgqGRtQ69iW5OxPOVSKR5IaIkgKeQAFjBpMWkf8MivALXy\n" +
  "8Ngyws0sazk2oLtlLEfaOHzBsSufdMPalAZI2T5XqcRs3ocSd5/F4sH/FIDDcRct\n" +
  "GL73N/4MJxwlbpmfCStnZYIHed6+iyysg3LccHkCgYBTPy0YejvOaTr3FMqIeL4r\n" +
  "3TsSbN9Q3/BX0oEvgTiJvLU1wsT3G0VzBE9qDbAmn9NuJhDKKgQRA7BdPVijZwib\n" +
  "NS49DtF2a5rB2wXfWNTG4KsQUPQ+zkhu3TyFUQijdlD+Vh47vD3bnH8hgInllUNc\n" +
  "WWPRkIP/TR8122xeYnv/hQ==\n" +
  "-----END PRIVATE KEY-----\n";
