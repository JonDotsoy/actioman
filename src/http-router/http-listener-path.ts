export const HTTP_LISTENER_PATH = new URL(
  "./http-listener.js",
  import.meta.url,
);
