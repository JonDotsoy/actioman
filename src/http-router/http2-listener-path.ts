export const HTTP2_LISTENER_PATH = new URL(
  "./http2-listener.js",
  import.meta.url,
);
