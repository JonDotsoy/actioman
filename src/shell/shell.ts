import { spawn, spawnSync } from "child_process";

type ShellConstructorOutput = "codeStatus" | "text" | "pipe";

type ShellConstructorOptions = {
  cmd: string;
  cwd?: string;
  signal?: AbortSignal;
  env?: Record<string, string>;
  output: ShellConstructorOutput;
  silent?: boolean;
};

export class ShellConstructor<T> {
  private cmd: string;
  private cwdOptions?: string;
  private abortController: AbortController;
  private envOptions: Record<string, string> | undefined;
  private appendEnvsOptions: Record<string, string> = {};
  private shouldThrow = false;
  private status: "pending" | "running" | "fulfilled" | "rejected" = "pending";
  private _promise = Promise.withResolvers<any>();
  private output: ShellConstructorOutput;
  private silent: boolean | undefined;
  private stdoutReadableController?: ReadableStreamDefaultController<Uint8Array>;
  private stdoutReadable = new ReadableStream<Uint8Array>({
    start: (controller) => {
      this.stdoutReadableController = controller;
    },
  });
  private stdoutWritable = new WritableStream<Uint8Array>({
    write: (chunk) => {
      if (!this.silent) process.stdout.write(chunk);
      if (this.output === "text" || this.output === "pipe")
        this.stdoutReadableController?.enqueue(chunk);
    },
    close: () => {
      this.stdoutReadableController?.close();
    },
  });
  private stderrWritable = new WritableStream<Uint8Array>({
    write: (chunk) => {
      if (!this.silent) process.stderr.write(chunk);
    },
  });
  runned = Promise.withResolvers<{ pid: number | undefined }>();

  constructor(options: ShellConstructorOptions) {
    this.abortController = new AbortController();
    this.cmd = options.cmd;
    this.cwdOptions = options.cwd;
    this.envOptions = options.env;
    this.output = options.output;
    this.silent = options.silent;
    options.signal?.addEventListener("abort", (event) => {
      this.abortController.abort();
    });
  }

  cwd(newCwd: string) {
    this.cwdOptions = newCwd;
    return this;
  }

  appendEnvs(env: Record<string, string>) {
    this.appendEnvsOptions = {
      ...this.appendEnvsOptions,
      ...env,
    };
    return this;
  }

  env(newEnv: Record<string, string>) {
    this.envOptions = newEnv;
    return this;
  }

  throws(shouldThrow: boolean = true) {
    this.shouldThrow = shouldThrow;
    return this;
  }

  quiet(silent: boolean = true) {
    this.silent = silent;
    return this;
  }

  text(): ShellConstructor<string> {
    this.output = "text";
    return this.quiet() as any;
  }

  background() {
    this.output = "pipe";
    const pid = this.runned.promise.then((r) => r.pid);

    let closed = false;
    return {
      pid,
      exited: this.then(),
      stdout: this.stdoutReadable,
      stderr: this.stderrWritable,
      close: async () => {
        if (closed) return;
        this.abortController.abort();
        const p = await pid;
        try {
          spawnSync(`kill ${p}`);
        } catch {}
        closed = true;
      },
    };
  }

  async run() {
    const cmd = this.cmd;
    const cwd = this.cwdOptions;
    const env = {
      ...(this.envOptions ?? process.env),
      ...this.appendEnvsOptions,
    };
    const abortController = this.abortController;
    const throws = this.shouldThrow;
    const output = this.output;
    const stdoutWritableWriter = this.stdoutWritable.getWriter();
    const stderrWritableWriter = this.stderrWritable.getWriter();

    const subprocess = spawn("sh", ["-c", cmd], {
      signal: abortController.signal,
      stdio: ["pipe", "pipe", "pipe"],
      env: env,
      cwd: cwd,
    });

    subprocess.stdout.addListener("data", (chunk: Uint8Array) => {
      stdoutWritableWriter.write(chunk);
    });
    subprocess.stdout.addListener("close", () => {
      stdoutWritableWriter.close();
    });

    subprocess.stderr.addListener("data", (chunk: Uint8Array) => {
      stderrWritableWriter.write(chunk);
    });
    subprocess.stderr.addListener("close", () => {
      stderrWritableWriter.close();
    });

    const formatOutput = async (exitCode: number) => {
      if (output === "pipe") return exitCode;
      if (output === "codeStatus") return exitCode;
      if (output === "text") {
        const text: number[] = [];
        const reader = await this.stdoutReadable.getReader();
        let readableStreamDefaultReadResult: Bun.ReadableStreamDefaultReadResult<Uint8Array>;
        while ((readableStreamDefaultReadResult = await reader.read())) {
          if (readableStreamDefaultReadResult.done) break;
          text.push(...readableStreamDefaultReadResult.value);
        }
        return new TextDecoder().decode(new Uint8Array(text));
      }
    };

    subprocess.addListener("error", (err) => {
      this._promise.reject(err);
    });

    subprocess.addListener("close", (exitCode) => {
      if (typeof exitCode === "number") {
        if (exitCode !== 0 && !throws)
          return this._promise.reject(
            new Error(`exitCode: ${exitCode} cmd: ${cmd}`),
          );
        return this._promise.resolve(formatOutput(exitCode));
      }

      return this._promise.reject(
        new Error(`exitCode is not a number: ${exitCode}`),
      );
    });
  }

  then<TResult1 = T, TResult2 = never>(
    onfulfilled?:
      | ((value: T) => TResult1 | PromiseLike<TResult1>)
      | undefined
      | null,
    onrejected?:
      | ((reason: any) => TResult2 | PromiseLike<TResult2>)
      | undefined
      | null,
  ): Promise<TResult1 | TResult2> {
    if (this.status === "pending") this.run();
    return this._promise.promise.then(onfulfilled, onrejected);
  }
}

export const $ = (
  arr: TemplateStringsArray,
  ...args: string[]
): ShellConstructor<number> => {
  return new ShellConstructor({
    cmd: String.raw(arr, ...args),
    output: "codeStatus",
  });
};

$.cwd =
  (cwd: string) =>
  (arr: TemplateStringsArray, ...args: string[]) =>
    $(arr, ...args).cwd(cwd);
$.env =
  (env: Record<string, string>) =>
  (arr: TemplateStringsArray, ...args: string[]) =>
    $(arr, ...args).env(env);
