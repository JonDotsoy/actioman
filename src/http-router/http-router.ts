import { Router } from "artur";
import { Actions, type Action } from "../actions/actions.js";
import { z } from "zod";
import { get } from "@jondotsoy/utils-js/get";
import { result } from "@jondotsoy/utils-js/result";
import { actionsToJson } from "../exporter-actions/exporter-actions.js";
import { Configs, type ConfigsModule } from "../configs/configs.js";
import { MetricsClient } from "../metric/metrics-client/metrics-client.js";

const isParser = (value: any): value is { parse: (value: any) => any } =>
  get.function(value, "parse") !== undefined;
const isZodType = (value: any): value is z.ZodType =>
  value instanceof z.ZodType;
const isSSEAction = (
  action: Action<any, any, any>,
): action is Action<any, any, true> => action.sse === true;

class EventStreamDataEncoder {
  encode(data: { id?: string; event?: string; data: any }) {
    const l = [];
    if (data.id) l.push(`id: ${data.id}`);
    if (data.event) l.push(`event: ${data.event}`);
    l.push(`data: ${JSON.stringify(data.data)}`);
    return new TextEncoder().encode(l.join("\n") + "\n\n");
  }
}

export class HTTPRouter {
  router: Router<"pass">;
  metrics: MetricsClient;

  constructor(
    actions: Actions,
    readonly configs?: Configs,
  ) {
    const actionsJson = actionsToJson(actions);

    this.metrics = new MetricsClient();

    this.router = new Router({
      errorHandling: "pass",
      middlewares: [
        ...(configs?.httpListenerMiddlewares ?? []),
        (fetch) => async (req) => {
          const res = await fetch(req);
          res?.headers.set("X-Powered-By", "Actioman");
          res?.headers.set("X-Actioman-Version", "v1.0.0");
          return res;
        },
      ],
    });

    configs?.httpSetup(this);

    this.router.use("GET", `/__actions`, {
      fetch: () => Response.json({ actions: actionsJson }),
    });

    for (const [name, describe] of Object.entries(Actions.describe(actions))) {
      const metricLabels = {
        action: name,
      };

      const requestErrorCountersMetric = this.metrics.counter({
        name: "request_error_counters",
        labels: metricLabels,
      });
      const requestCountersMetric = this.metrics.counter({
        name: "request_counters",
        labels: metricLabels,
      });
      const requestDataTransferenceBytesMetric = this.metrics.counter({
        name: "request_data_transference_bytes",
        labels: metricLabels,
      });
      const requestDurationSecondsMetric = this.metrics.averageBySecond({
        name: "request_duration_seconds",
        labels: metricLabels,
      });

      const isSSE = isSSEAction(describe);
      const method = "POST";
      this.router.use(method, `/__actions/${name}`, {
        middlewares: [
          (fetch) => (req) => {
            const start = Date.now();
            requestCountersMetric.increment();
            return Promise.resolve(fetch(req))
              .catch((err) => {
                requestErrorCountersMetric.increment();
                throw err;
              })
              .finally(() => {
                requestDurationSecondsMetric.add(Date.now() - start);
              });
          },
        ],
        fetch: async (req) => {
          if (isSSE) return this.sseHandler(describe, req);
          return this.handler(describe, req);
        },
      });
    }
  }

  private async sseHandler(action: Action<any, any, true>, req: Request) {
    const url = new URL(req.url);
    const inputSearchParams = url.searchParams.get("input") ?? null;
    const [, data] = await result(() =>
      inputSearchParams ? JSON.parse(inputSearchParams) : null,
    );
    const input = isParser(action.input) ? action.input.parse(data) : null;
    const resultAction = await action.handler(input);
    const body = new ReadableStream<Uint8Array>({
      pull: async (ctrl) => {
        const [err, iteratorResult] = await result(() => resultAction.next());

        if (err) {
          console.error(err);
          ctrl.enqueue(
            new EventStreamDataEncoder().encode({
              event: "error",
              data: "Internal error",
            }),
          );
          return ctrl.close();
        }

        const { done, value } = iteratorResult;

        if (done) {
          ctrl.enqueue(
            new EventStreamDataEncoder().encode({
              event: "close",
              data: true,
            }),
          );
          return ctrl.close();
        }

        const output = isParser(action.output)
          ? action.output.parse(value)
          : value;

        ctrl.enqueue(new EventStreamDataEncoder().encode({ data: output }));
      },
    });
    return new Response(body, {
      headers: {
        "Content-Type": "text/event-stream",
        "Cache-Control": "no-cache",
        Connection: "keep-alive",
        "X-Accel-Buffering": "no",
      },
    });
  }

  private async handler(action: Action<any, any, any>, req: Request) {
    const [, data] = await result(() => req.json());
    const input = isParser(action.input) ? action.input.parse(data) : null;
    const resultDescribe = await action.handler(input);
    const output = isParser(action.output)
      ? action.output.parse(resultDescribe)
      : resultDescribe;
    return Response.json(output);
  }

  static fromModule(module: unknown, configs?: ConfigsModule) {
    const actions = Actions.fromModule(module, configs);
    if (!actions) throw new Error(`No actions found in ${module}`);
    return new HTTPRouter(actions, Configs.fromModule(configs));
  }
}
