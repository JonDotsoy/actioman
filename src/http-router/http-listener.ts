import * as http from "http";
import { HTTPRouter } from "./http-router.js";
import net from "net";
import type { ConfigsModule } from "../configs/configs.js";

type ListenOptions = {
  silent: boolean;
};

const sanatizeHostname = (hostname: string) => {
  switch (hostname) {
    case "::1":
    case "::":
    case "0.0.0.0":
    case "127.0.0.1":
    case "::ffff:127.0.0.1":
      return "localhost";
  }

  return hostname;
};

const INITIAL_PORT = 30_320;

const findNextPort = async () => {
  let porposalPort = INITIAL_PORT;
  while (true) {
    porposalPort++;
    if (porposalPort >= INITIAL_PORT + 10_000)
      throw new Error("No available port");
    const port = await new Promise<number | null>((resolve) => {
      const connectiong = net.connect({
        host: "localhost",
        port: porposalPort,
      });

      connectiong.addListener("connect", () => {
        resolve(null);
        connectiong.destroy();
      });
      connectiong.addListener("error", (err) => {
        if ("code" in err && err.code === "ECONNREFUSED") resolve(porposalPort);
      });
    });
    if (typeof port === "number") return port;
  }
};

export class HTTPLister {
  server: http.Server;

  constructor(readonly httpRouter: HTTPRouter) {
    this.server = http.createServer(async (req, res) => {
      try {
        if (!(await httpRouter.router.requestListener(req, res))) {
          res.statusCode = 404;
          return res.end();
        }
      } catch (error) {
        console.error(error);
        res.statusCode = 500;
        return res.end();
      }
    });
  }

  async close() {
    await new Promise<null>((resolve, reject) => {
      if (this.server.address() === null) return resolve(null);
      this.server.close((err) => {
        if (err) return reject(err);
        return resolve(null);
      });
    });
  }

  async listen(port?: number, hostname?: string, options?: ListenOptions) {
    const silent = options?.silent ?? HTTPLister.defaultOptions.silent;

    const log = (message: string) => {
      if (silent) return;
      console.log(message);
    };

    const portToListen = port ?? (await findNextPort());
    const hostnameToListen = hostname ?? "localhost";

    const url = await new Promise<URL>((resolve) => {
      this.server.listen(portToListen, hostnameToListen, () => {
        const address = this.server.address();
        if (typeof address === "object" && address !== null)
          return resolve(
            new URL(
              `http://${sanatizeHostname(address.address)}:${address.port}`,
            ),
          );
        return resolve(
          new URL(
            `http://${sanatizeHostname(hostnameToListen)}:${portToListen}`,
          ),
        );
      });
    });

    if (!silent) {
      for (const route of this.httpRouter.router.routes) {
        const method = route.method;
        const urlPattern = route.urlPattern;
        log(`Route ${method} ${urlPattern.pathname}`);
      }
    }
    log(`Listening on ${url.toString()}`);

    return url;
  }

  static fromModule(module: unknown, configs?: ConfigsModule) {
    const httpRouter = HTTPRouter.fromModule(module, configs);
    if (!httpRouter) throw new Error("No actions found");
    return new HTTPLister(httpRouter);
  }

  static defaultOptions = {
    silent: false,
  };
}
