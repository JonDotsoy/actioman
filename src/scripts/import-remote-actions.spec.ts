import { afterEach, beforeAll, beforeEach, describe, it } from "bun:test";
import { $ } from "../shell/shell";
import { importRemoteActions } from "./import-remote-actions";
import { HTTPLister } from "../http-router/http-listener";
import { defineAction } from "../actions/actions";
import { set, z } from "zod";

describe("import-remote-actions", () => {
  const cleanupTasks = new Set<() => any>();

  afterEach(async () => {
    for (const cleanupTask of cleanupTasks) await cleanupTask();
    cleanupTasks.clear();
  });

  beforeAll(async () => {
    await $`
      rm -rf __tests__/tmp/
      echo "*" > __tests__/.gitignore
      mkdir -p __tests__/tmp/
      mkdir -p __tests__/tmp/node_modules/actioman/lib/esm/share-actions/
      mkdir -p __tests__/tmp/node_modules/actioman/lib/esm/actions-target/
      echo 'export * from "../../../../../../../../actions-target/actions-target.js";' > __tests__/tmp/node_modules/actioman/lib/esm/actions-target/actions-target.js
      echo "export const shareActions = {};" > __tests__/tmp/node_modules/actioman/lib/esm/share-actions/share-actions.js
    `.cwd(new URL("./", import.meta.url).pathname);
  });

  it("imports remote actions", async () => {
    const httpLister = HTTPLister.fromModule({
      hi: defineAction({
        input: z.object({ name: z.string() }),
        handler: async ({ name }) => `Hello, ${name}!`,
      }),
    });
    cleanupTasks.add(() => httpLister.close());

    const server = await httpLister.listen(0);

    await importRemoteActions(
      server,
      "1.- my actions 1",
      new URL("./__tests__/tmp/", import.meta.url).pathname,
    );
  });
});
