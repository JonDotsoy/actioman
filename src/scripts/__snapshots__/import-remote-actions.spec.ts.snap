// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`import-remote-actions imports remote actions: file content 1`] = `
"// @ts-nocheck
import { z } from "zod";
import { ActionsTarget } from "./../../actions-target/actions-target.js";
const createActionsTarget = () => new ActionsTarget("http://localhost:43551/__actions", {
    hi: {
        description: undefined,
        input: z.object({ "name": z.string() }).strict(),
        output: undefined
    }
}).compile();
export default createActionsTarget;
"
`;

exports[`import-remote-actions imports remote actions: file content 2`] = `
"export {
    default as _1MyActions1
} from './remote_actions/_1MyActions1.js';"
`;
